### PR TITLE
Rewrite BBDuk single-end and BBDuk paired-end processes to Python

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist
* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [ ] All output fields have a value assigned to them.

### CODE REVIEW

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 1]
```diff
@@ -0,0 +1,1223 @@
...
+        statistics = ListField(FileField(), label="Statistics")
+        fastqc_url = ListField(FileHtmlField(), label="Quality control with FastQC")
+        fastqc_archive = ListField(FileField(), label="Download FastQC archive")
+
+    def run(self, inputs, outputs):
+        """Run analysis."""
+        reads_basename = Path(inputs.reads.output.fastq[0].path).name
+        assert reads_basename.endswith(".fastq.gz")
+        name = reads_basename[:-9]
+        input_reads = "input_reads.fastq.gz"
```
Comment (romunov) [inline resolwe_bio/processes/reads_processing/bbduk.py:448]: Add new line above to make "inputs" visually be distringuished from the "name" chunk of code.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 2]
```diff
@@ -0,0 +1,1223 @@
...
+        )()
+        if inputs.reference.sequences:
+            with open(input_references, "wb") as outfile:
+                for sequences in inputs.reference.sequences:
+                    with open(sequences.output.fastagz.path, "rb") as infile:
+                        shutil.copyfileobj(infile, outfile)
+
+        if inputs.header_parsing.barcode_files:
+            with open(input_barcodes, "wb") as outfile:
+                for barcode_files in inputs.header_parsing.barcode_files:
```
Comment (romunov) [inline resolwe_bio/processes/reads_processing/bbduk.py:466]: `barcode_files` could be singular (so `for barcode_file in ...`) to indicate that it's iterating through a list of files.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 3]
```diff
@@ -0,0 +1,1223 @@
...
+        input_barcodes = "input_barcodes.fasta.gz"
+        final_reads = name + "_preprocessed.fastq"
+        final_statistics = name + "_statistics.txt"
+
+        (
+            Cmd["cat"][[reads.path for reads in inputs.reads.output.fastq]]
+            > input_reads
+        )()
+        if inputs.reference.sequences:
+            with open(input_references, "wb") as outfile:
```
Comment (romunov) [inline resolwe_bio/processes/reads_processing/bbduk.py:459]: Where is this used?

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 4]
```diff
@@ -0,0 +1,1223 @@
...
+
+        args_fastqc = [
+            f"{final_reads}.gz",
+            "fastqc",
+            "fastqc_archive",
+            "fastqc_url",
+        ]
+
+        if inputs.fastqc.nogroup:
+            args.append("--nogroup")
```
Comment (romunov) [inline resolwe_bio/processes/reads_processing/bbduk.py:575]: Probably `args_fastqc.append()`?

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 5]
```diff
@@ -0,0 +1,1223 @@
...
+
+    class Output:
+        """Output fields."""
+
+        fastq = ListField(FileField(), label="Remaining reads")
+        statistics = ListField(FileField(), label="Statistics")
+        fastqc_url = ListField(FileHtmlField(), label="Quality control with FastQC")
+        fastqc_archive = ListField(FileField(), label="Download FastQC archive")
+
+    def run(self, inputs, outputs):
```
Comment (romunov) [inline resolwe_bio/processes/reads_processing/bbduk.py:443]: Consider adding `self.progress()` throughout the process to show progress where this functionality may be used.

**Review by romunov — COMMENTED** [review]
Had a chance to look through the SingleEnd, will work on the rest next week.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 6]
```diff
@@ -0,0 +1,1223 @@
...
+    quality-score recalibration, kmer cardinality estimation, and various
+    other operations in a single pass. See
+    [here](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/)
+    for more information.
+    """
+
+    slug = "bbduk-single"
+    name = "BBDuk (single-end)"
+    process_type = "data:reads:fastq:single:bbduk"
+    version = "2.6.0"
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk.py:41]: I was also wondering about this but there are no breaking changes.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 7]
```diff
@@ -0,0 +1,1223 @@
...
+            )
+
+        class Processing:
+            """Processing parameters."""
+
+            kmer_length = IntegerField(
+                label="Kmer length",
+                default=27,
+                description="Kmer length used for finding contaminants. "
+                "Contaminants shorter than Kmer length will not be found. "
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk.py:101]: Agree. Will change "Kmer" to lowercase unless "Kmer" is the first word of the sentence.

File: resolwe_bio/processes/reads_processing/bbduk.py (hunk 3 above)
Comment (jkokosar) [reply resolwe_bio/processes/reads_processing/bbduk.py:472]: This prepares a single "adapters" gzip file from a possible list of multiple input .fasta files containing adapters to be removed.

@OtonicarJan the line should be `with gzip.open(input_references, "wb") as outfile:` so don't forget to import the `gzip` module and fix this in both processes.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 8]
```diff
@@ -0,0 +1,1275 @@
...
+    def run(self, inputs, outputs):
+        """Run analysis."""
+        reads_fw_basename = Path(inputs.reads.output.fastq[0].path).name
+        assert reads_fw_basename.endswith(".fastq.gz")
+        name_fw = reads_fw_basename[:-9]
+        reads_rv_basename = Path(inputs.reads.output.fastq2[0].path).name
+        assert reads_rv_basename.endswith(".fastq.gz")
+        name_rv = reads_rv_basename[:-9]
+
+        input_fw_reads = "input_fw_reads.fastq.gz"
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/bbduk.py:1092]: The naming of the reads and variables was not optimal in the original process. Please use mate1/mate2 terminology instead, and replace the relevant lines, for example `input_fw_reads = "input_fw_reads.fastq.gz"` could be  `input_mate1_reads = "input_mate1_reads.fastq.gz"`

**Review by jkokosar — COMMENTED** [review]
Good job! I only have one terminology suggestion left. Please see the review comment.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 9]
```diff
@@ -0,0 +1,1223 @@
...
+            )
+            hamming_distance = IntegerField(
+                label="Maximum Hamming distance for kmers (substitutions only)",
+                default=0,
+            )
+            query_hamming_distance = IntegerField(
+                label="Hamming distance for query kmers",
+                default=0,
+            )
+            edit_distance = IntegerField(
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk.py:148]: Couldn't find descriptions for parameters that have unresolved comments.

File: resolwe_bio/processes/reads_processing/bbduk.py [hunk 10]
```diff
@@ -0,0 +1,1223 @@
...
+            )
+
+        class Complexity:
+            """Complexity parameters."""
+
+            entropy = FloatField(
+                label="Minimum entropy",
+                default=-1.0,
+                description="Set between 0 and 1 to filter reads with entropy below that value. "
+                "Higher is more stringent.",
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk.py:394]: -1 is the default value for the tool. I think this means that this parameter is off.

**Review by romunov — APPROVED** [review]

**Pull request merged**